File: leetcode/python/0104_maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root == None: return 0
        if root.left == None and root.right == None: return 1
        
        l = self.maxDepth(root.left)
        r = self.maxDepth(root.right)
        return max(l, r) + 1 

File: leetcode/cpp/0104_maximum_depth_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepthHelper(TreeNode* root, int& max, int& currMax) {
        if (root != nullptr) {
            currMax++;  

            maxDepthHelper(root->left, max, currMax);

            if (max < currMax) {
                max = currMax;
            }

            maxDepthHelper(root->right, max, currMax);
            currMax--;
        }

        return max;
    }
    
    int maxDepth(TreeNode* root) {
        int max = 0, currMax = 0;

        return maxDepthHelper(root, max, currMax);
    }
};